Civi/Api4/Generic/Action/AbstractBatch.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Civi\Api4\Generic\Action;
+
+/**
+ * Base class for all batch actions (Update, Delete, Replace).
+ *
+ * This differs from the AbstractQuery class in that the "Where" clause is required.
+ *
+ * @package Civi\Api4\Generic
+ */
+abstract class AbstractBatch extends AbstractQuery {
+
+  /**
+   * Criteria for selecting items to process.
+   *
+   * @required
+   * @var array
+   */
+  protected $where = [];
+
+  /**
+   * @var array
+   */
+  private $select;
+
+  public function __construct($entity = NULL, $idField = 'id') {
+    $this->select = (array) $idField;
+    parent::__construct($entity);
+  }
+
+  /**
+   * @return array
+   */
+  protected function getBatchItems() {
+    $params = $this->getParams();
+    if (empty($this->reload)) {
+      $params['select'] = $this->getSelect();
+    }
+
+    $action = civicrm_api4($this->getEntity(), 'getActions', ['where' => [['name', '=', 'get']], 'select' => ['params']])->first();
+    $params = array_intersect_key($params, $action['params']);
```

Comment:
This took me a couple minutes to get, but it's neat. 👍 

My draft had been more complicated (`BasicBulkTask` had required both `$getter` and `$doer`). The simplification delegates to the `get` action instead of requiring an explicit `$getter` -- and that *feels* pretty right. It seems valid in all the examples I've had, and (thinking about how different UX's evolve on top of APIs) it seems like a generally wise prescription to ensure that all batch-tasks have a corresponding "get" (which can generate a *preview* of what the batch-tasks will do).

Regarding which properties to proapagate...  The `array_intersect_key(...)` seemed clever at first, but I'm starting to get apprehensive about it. Trying to think about this concretely:

* __Definitely propagate__: `checkPermissions`, `where`
* __Probably propagate__: `orderBy`, `limit`, `offset`
* __Unknown__: Bespoke parameters (for the specific get and the specific task) which happen to have the same name.

The validity of propagating through `checkPermissions`, `where`, `orderBy`, `limit`, and `offset` arises from the fact that `AbstractBatch` and `Get` have common parents (`AbstractQuery`, `AbstractAction`). So those bits are good.

I can't reason about unknown params cleanly (because... they're unknown...), but here's one to consider: `AbstractEntity.php` has a comment which anticipates a new parameter named `$chain` (which I'm guessing is some list of additional actions to perform). When/if that property comes online, I'm pretty sure we would *not* want it copied over.

My gut says to use a simple/static whitelist here.

Reply:
You make some good points there. I didn't want to box us in incase a custom api's Get class has some important extra param that must be passed through. But you're right it opens us up to weird stuff bleeding through as well.
Considering that this function itself could be overridden by some highly customized api if they really need it to do something different, yea I agree we're better off with a whitelist.